flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/SparseVector.java

```diff
@@ -0,0 +1,791 @@
+	 */
+	public SparseVector(int n, Map<Integer, Double> kv) {
+		this.n = n;
+		int nnz = kv.size();
+		int[] indices = new int[nnz];
+		double[] values = new double[nnz];
+
+		int pos = 0;
+		for (Map.Entry<Integer, Double> entry : kv.entrySet()) {
+			indices[pos] = entry.getKey();
+			values[pos] = entry.getValue();
+			pos++;
+		}
+
+		this.indices = indices;
+		this.values = values;
+		checkIndices();
+
+		if (!(kv instanceof TreeMap)) {
+			sortIndices();
+		}
+	}
+
+	private void checkIndices() {
+		if (indices.length != values.length) {
+			throw new IllegalArgumentException("Indices size and values size should be the same.");
+		}
+		for (int i = 0; i < indices.length; i++) {
+			if (indices[i] < 0 || (n >= 0 && indices[i] >= n)) {
+				throw new IllegalArgumentException("Index out of bound.");
+			}
+		}
+	}
+
+	private void sortIndices() {
+		boolean outOfOrder = false;
+		for (int i = 0; i < this.indices.length - 1; i++) {
+			if (this.indices[i] >= this.indices[i + 1]) {
+				outOfOrder = true;
+				break;
+			}
+		}
+
+		if (!outOfOrder) {
+			return;
+		}
+
+		// sort
+		Integer[] order = new Integer[this.indices.length];
+		for (int i = 0; i < order.length; i++) {
+			order[i] = i;
+		}
+
+		Arrays.sort(order, new Comparator<Integer>() {
+			@Override
+			public int compare(Integer o1, Integer o2) {
+				if (indices[o1] < indices[o2]) {
+					return -1;
+				} else if (indices[o1] > indices[o2]) {
+					return 1;
+				} else {
+					return 0;
+				}
+			}
+		});
+
+		int nnz = this.indices.length;
+		int[] sortedIndices = new int[nnz];
+		double[] sortedValues = new double[nnz];
+
+		for (int i = 0; i < order.length; i++) {
+			sortedValues[i] = this.values[order[i]];
+			sortedIndices[i] = this.indices[order[i]];
+		}
+
+		this.indices = sortedIndices;
+		this.values = sortedValues;
+	}
+
+	@Override
+	public SparseVector clone() {
+		SparseVector vec = new SparseVector(this.n);
+		vec.indices = this.indices.clone();
+		vec.values = this.values.clone();
+		return vec;
+	}
+
+	@Override
+	public SparseVector prefix(double d) {
+		int[] indices = new int[this.indices.length + 1];
+		double[] values = new double[this.values.length + 1];
+		int n = (this.n >= 0) ? this.n + 1 : this.n;
+
+		indices[0] = 0;
+		values[0] = d;
+
+		for (int i = 0; i < this.indices.length; i++) {
+			indices[i + 1] = this.indices[i] + 1;
+			values[i + 1] = this.values[i];
+		}
+
+		return new SparseVector(n, indices, values);
+	}
+
+	@Override
+	public SparseVector append(double d) {
+		int[] indices = new int[this.indices.length + 1];
+		double[] values = new double[this.values.length + 1];
+		int n = (this.n >= 0) ? this.n + 1 : this.n;
+
+		int i;
+		for (i = 0; i < this.indices.length; i++) {
+			indices[i] = this.indices[i];
+			values[i] = this.values[i];
+		}
+
+		indices[i] = n - 1;
+		values[i] = d;
+
+		return new SparseVector(n, indices, values);
+	}
+
+	public int[] getIndices() {
```

Comment:
maybe explain why these only exist in `SparseVector`

Reply:
Thanks for the suggestion. But in my opinion, using an indices array and a values array to represent a sparse vector is the convention in machine learning as well as other domains like high performance computing.